### PR TITLE
discourse: 2.7.9 -> 2.8.0.beta9

### LINKED ISSUE #146308: Should we package Discourse beta in nixpkgs?
Currently we package the Discourse Stable branch in Nixpkgs. Some questions I'd like to discuss:

1. Should we instead only package the Beta branch?
2. Or, should we package both? (If yes, should we maintain parallel plugins for both branches?)
3. Or, should we keep only packaging Stable?

## Some background

* An official description of the Discourse branches: https://meta.discourse.org/t/updates-always-come-before-release-notes/88919/7
* Upstream recommends following the Beta branch as a good default https://meta.discourse.org/t/please-dont-pressure-self-installers-to-be-on-beta-branch/32237/2?u=ryantm
  > You are free to choose what’s right for you, but we recommend that people follow the betas unless they are extremely risk averse.
* Beta is the default branch you get when you install Discourse from with Docker.
* All the professionally hosted (by upstream) instances of Discourse follow the Beta branch.
* https://discourse.nixos.org is using Discourse Beta branch. 
* I am also using Discourse Beta branch in my deployments.

## Discussion

I think we should focus our efforts on only packaging the Beta branch to more closely follow upstream and what we are deploying. I'd be in favor of having instructions for how someone could package the Stable branch themselves, but I wouldn't want to keep maintaining that version personally.

cc @talyz @dpausp

### LINKED ISSUE #147867: Vulnerability roundup 108: discourse-2.7.9: 2 advisories [7.5]
[search](https://search.nix.gsc.io/?q=discourse&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=discourse+in%3Apath&type=Code)

* [x] [CVE-2021-41082](https://nvd.nist.gov/vuln/detail/CVE-2021-41082) CVSSv3=7.5 (nixos-21.05, nixos-unstable)
* [ ] [CVE-2021-41271](https://nvd.nist.gov/vuln/detail/CVE-2021-41271) CVSSv3=5.3 (nixos-21.05, nixos-unstable)

### CVE-2021-41082

Discourse is a platform for community discussion. In affected versions any private message that includes a group had its title and participating user exposed to users that do not have access to the private messages. However, access control for the private messages was not compromised as users were not able to view the posts in the leaked private message despite seeing it in their inbox. The problematic commit was reverted around 32 minutes after it was made. Users are encouraged to upgrade to the latest commit if they are running Discourse against the `tests-passed` branch.

### CVE-2021-41271

Discourse is a platform for community discussion. In affected versions a maliciously crafted request could cause an error response to be cached by intermediate proxies. This could cause a loss of confidentiality for some content. This issue is patched in the latest stable, beta and tests-passed versions of Discourse.

-----
Scanned versions: nixos-21.05: 4f37689c8a2; nixos-unstable: 942eb9a335b.

Cc @talyz

### PR DESCRIPTION
**Pull request by talyz**
###### Motivation for this change
- Update Discourse to the latest beta release, as discussed in #146308
- Add a `libv8` output to `nodejs` as an alternative to `v8`, which is a lot of work to update
- Update Discourse plugins to the latest releases and make it possible to update `discourse-prometheus`
- Replace `up-plugins.sh` with a `README` pointing to the plugin packaging docs
- Improve version handling in the `update.py` script
- Implement Discourse's [plugin pinning algorithm](https://meta.discourse.org/t/pinning-plugin-and-theme-versions-for-older-discourse-installs/156971) in the `update.py` script

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by ryantm — APPROVED** [review]
Code changes look good, as far as I can assess. I don't know much about the v8 changes.

I've upgraded a few of my deployments with it without trouble. Tested the following plugins work: LDAP, data-explorer, solved, checklist, calendar.

File: nixos/modules/services/web-apps/discourse.nix [hunk 1]
```diff
@@ -824,8 +827,14 @@ in
       upstreams.discourse.servers."unix:/run/discourse/sockets/unicorn.sock" = {};
 
       appendHttpConfig = ''
+        # Additional MIME types that nginx should handle
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/discourse.nix:830]: Our nginx uses mailcap mime.types which already includes those.

File: pkgs/development/web/nodejs/nodejs.nix [hunk 2]
```diff
@@ -130,6 +134,35 @@ let
 
       # install the missing headers for node-gyp
       cp -r ${concatStringsSep " " copyLibHeaders} $out/include/node
+
+      # assemble a static v8 library and put it in the 'libv8' output
+      mkdir -p $libv8/lib
+      pushd out/Release/obj.target
+      find . -path "./torque_*/**/*.o" -or -path "./v8*/**/*.o" | sort | uniq >files
```
Comment (SuperSandro2000) [inline pkgs/development/web/nodejs/nodejs.nix:141]:
```suggestion
      find . -path "./torque_*/**/*.o" -or -path "./v8*/**/*.o" | sort -u >files
```
sort can do that and maybe it is even faster for a lot of files.

File: pkgs/servers/web-apps/discourse/plugins/discourse-prometheus/prometheus_exporter_version [hunk 3]
```diff
@@ -0,0 +1 @@
+0.5.0
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/discourse/plugins/discourse-prometheus/prometheus_exporter_version:1]: It is not very obvious why we need this file. Could it be also generated on the fly?

Comment (talyz) [reply pkgs/servers/web-apps/discourse/plugins/discourse-prometheus/prometheus_exporter_version:1]: It's imported from the plugin's repo by the `update.py` script and read by the `Gemfile`.

File: pkgs/development/web/nodejs/nodejs.nix (hunk 2 above)
Comment (talyz) [reply pkgs/development/web/nodejs/nodejs.nix:141]: Yeah, good point. It's probably a bit faster.

File: nixos/modules/services/web-apps/discourse.nix (hunk 1 above)
Comment (talyz) [reply nixos/modules/services/web-apps/discourse.nix:830]: Ah, indeed we do. Removed. :)

**Comment by talyz** [comment]
I've updated to `2.8.0.beta9` and improved the version handling in the `update.py` script. It should now look for the latest release among _all_ available version numbered releases. It also now implements Discourse's [plugin pinning algorithm](https://meta.discourse.org/t/pinning-plugin-and-theme-versions-for-older-discourse-installs/156971) to make sure we only update plugins to versions which are compatible with our packaged version of Discourse. While this likely won't matter much most of the time if we continue packaging the beta versions, it could be helpful if we decide to go back to packaging
release versions or if we run into issues with future upgrades. In that case, the plugins could still be updated safely even though we're not on the latest version of Discourse.

**Comment by ryantm** [comment]
Cool, @talyz. I deployed your latest thing to my deployments and it seems to be working fine.

**Comment by talyz** [comment]
I've added a fix for the `/public` path issue from #142528 that I think is better than the current patch. I recently noticed that backup downloads don't work and it turned out that the root cause was the same issue (a mismatch between how nginx and Discourse refer to the same path in sendfile requests). There may be more places we haven't encountered yet, and even more could crop up in the future, so I've fixed the issue by updating the nginx configuration instead.

**Comment by ryantm** [comment]
@talyz that change looks good to me

**Comment by talyz** [comment]
I'll merge this on Wednesday (2021-12-08) unless anyone objects. It would be nice to get some more eyes on the v8 change, but it shouldn't affect nodejs (unless it fails the build), since it doesn't touch any of the default outputs.

**Review by dpausp — COMMENTED** [review]
Thanks! Code changes look good to me, upgraded my test system from 2.7.9. Tested basic posting functionality, backups and the following plugins: discourse-assign, discourse-canned-replies, discourse-docs, discourse-spoiler-alert, discourse-voting.

**Pull request merged**